### PR TITLE
New Witness Format (and remove the rest of the tapes)

### LINKED ISSUE #179: block witness: replace tapes with in-line operands
In order to minimise buffering requirements when producing and interpreting block witnesses, it is more convenient to provide all necessary arguments as operands to the instructions `LEAF`, `ACCOUNTLEAF`, `HASH`, etc.
It may make `GenStructStep` function more tricky, perhaps IT will need to have tapes instead.

Another benefit (and this is how the idea of this change came about initially) is the possibility of describing formal semantics of the block witness stack machine in a way similar to Web Assembly

This change will require changing the way `ACCOUNTLEAF` retrieves its arguments. Currently, two of them (nonce and balance) are expected to be provided with the tapes, and another two, Root and CodeHash, are supposed to be on the stack. And the `fieldSet` tells us which of the four present. We need to make it more similar to `BRANCH` instruction. All arguments would need to come from the stack (like in `BRANCH`), and `fieldSet` will be similar to `set` argument in `BRANCH`

### PR DESCRIPTION
**Pull request by mandrigin**
## Known Issue:
hash mismatch on block 5.000.000: https://github.com/ledgerwatch/turbo-geth/issues/316

~~account balance deserialization issue: #319~~ (fixed)

---
fixes #179

### CODE REVIEW

**Comment by mandrigin** [comment]
The hash issue is quite slow to debug, so I would prefer to merge that and then fix the issue. It shouldn't be a lot of changes.

File: trie/witness_operators.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+	if !ok {
+		return []byte{}, fmt.Errorf("unexpected decode result, expected []byte, got %T", buffer)
+	}
+
+	return b, nil
+}
+
+func encodeByteArray(data []byte, output io.Writer) error {
+	encoder := codec.NewEncoder(output, &cbor)
+	return encoder.Encode(data)
```
Comment (AlexeyAkhunov) [inline trie/witness_operators.go]: Perhaps use `encode.Encode(&data)` for better performance

File: trie/witness_operators.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+	b, ok := buffer.([]byte)
+	if !ok {
+		return []byte{}, fmt.Errorf("unexpected decode result, expected []byte, got %T", buffer)
+	}
+
+	return b, nil
+}
+
+func encodeByteArray(data []byte, output io.Writer) error {
+	encoder := codec.NewEncoder(output, &cbor)
```
Comment (AlexeyAkhunov) [inline trie/witness_operators.go]: creating encoder for each encoding is quite expensive. Need a pool or at least one encoder per witness

File: trie/witness_operators.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+		return err
+	}
+
+	o.Key = keyBytesToNibbles(key)
+
+	return nil
+}
+
+func decodeByteArray(input io.Reader) ([]byte, error) {
+	codec := codec.NewDecoder(input, &cbor)
```
Comment (AlexeyAkhunov) [inline trie/witness_operators.go]: Creating a decoder per operation is expensive. Need a pool or at least decoder per witness

File: trie/witness_operators.go (hunk 2 above)
Comment (mandrigin) [reply trie/witness_operators.go]: one per witness is a trivial change for this structure, will start with that...

File: trie/witness_operators.go (hunk 1 above)
Comment (mandrigin) [reply trie/witness_operators.go]: will it change much? the data in slices is already passed by reference, it is just the metadata that will be copied: https://play.golang.org/p/Y9kWdYdbXLx

File: trie/witness.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	return nil
+}
+
+func defaultWitnessHeader() WitnessHeader {
+	return WitnessHeader{WitnessVersion}
+}
+
+type Witness struct {
+	Header   WitnessHeader
+	Operands []WitnessOperator
```
Comment (AlexeyAkhunov) [inline trie/witness.go]: Change to `Operators`

File: trie/witness_operators.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+		flags |= flagCode
+	}
+	if o.HasStorage {
+		flags |= flagStorage
+	}
+	if o.Nonce > 0 {
+		flags |= flagNonce
+	}
+
+	emptyBalance := big.NewInt(0)
```
Comment (mandrigin) [inline trie/witness_operators.go]: replace to `Sign`

**Comment by mandrigin** [comment]
red tests because of an issue in master

**Pull request merged**